### ISSUE TITLE
Integration between event-loop of AudioWorkletGlobalScope and rendering loop

### ISSUE THREAD

**Issue by gterzian** [WebAudio/web-audio-api#2008]
**Describe the issue**

Web Audio describes a control-thread, and a rendering-thread, where the control-thread runs a "traditional" event-loop as described in HTML, and the rendering thread is running a custom "rendering loop", described at https://webaudio.github.io/web-audio-api/#rendering-loop. 

Step 3 of the rendering loop, "Process a render quantum.", ends-up, in the case of a ` AudioWorkletNode`, calling into the `process` method of the corresponding ` AudioWorkletProcessor`. 

It should be noted that the  `AudioWorkletProcessor` runs in a `AudioWorkletGlobalScope`, which is a sub-class of `WorkletGlobalScope`, which should have it's own distinct event-loop as described in https://drafts.css-houdini.org/worklets/#the-event-loop

A further complication is that each `AudioWorkletProcessor`  also has a shipped `MessagePort`, whose `post-message-queue` should be treated as a first-class task-queue for the "event-loop" on which it happens to be, in this case the event-loop of the `AudioWorkletGlobalScope`. 

So the problem is that it's not clear how the rendering loop integrates with the event-loop of the `AudioWorkletGlobalScope`. 

> The rendering thread has another task queue for microtasks for any microtask operation such as resolution of Promises in the AudioWorkletGlobalScope.

This seems to imply a kind of integration where the rendering loop would run a microtask checkpoint for the `AudioWorkletGlobalScope`, as is found in Step 4 of the rendering loop. 

However that doesn't cover messages received on the port, since those aren't microtasks. 

If we take this example:

https://github.com/GoogleChromeLabs/web-audio-samples/blob/master/audio-worklet/design-pattern/shared-buffer/shared-buffer-worklet-processor.js

We see an interplay between the `onmessage` handler on the port, and the `process` method, where `process` essentially does nothing until an "initialization" message has been received on the port. 

In practice, how does a UA interleave running tasks on the `AudioWorkletGlobalScope`, for example when a new message is received on the port, with running the rendering loop which itself calls into `process` of the processor running in the context of `AudioWorkletGlobalScope`?

And what about tasks enqueued on the event-loop of `AudioWorkletGlobalScope` as a result of the call to https://drafts.css-houdini.org/worklets/#dom-worklet-addmodule ?

I can imagine two ways to do it in practice with the current spec:  

1. The task-queues from the event-loop of `AudioWorkletGlobalScope` could be plugged into the "control-message-queue" of the rendering thread? 
2. run the event-loop of `AudioWorkletGlobalScope` in parallel to the render-loop, but somehow "stop" it when the render-loop wants to call `process`?

It appears pretty clear that the goal is not running the event-loop of the  `AudioWorkletGlobalScope` fully in parallel to the render loop, since that would create potential race-conditions between the `onmessage` handler of a processor and it's `process` method. Yet I cannot find anything in the spec that integrates both event-loops so as to interleave both sequentially in some way. 

A solution could be rewording https://webaudio.github.io/web-audio-api/#rendering-loop, where step 4 would process all task-queues of the event-loop of the `AudioWorkletGlobalScope`, with microtask checkpoints interleaved?

**Where Is It**
https://webaudio.github.io/web-audio-api/#rendering-thread

**Additional Information**
Could be relevant for worklets in general, see https://github.com/whatwg/html/issues/4213

**Comment by gterzian** [comment]
Seeing this was discussed in https://github.com/WebAudio/web-audio-api/issues/1511

To me, altough that previous issue was closed, the spec in it's current state is still not clear. I think an improvement was made by describing how microtasks are run, and that still leaves handling of proper tasks from the event-loop of `AudioWorkletGlobalScope`(for example from incoming message on the `MessagePort`) as an exercise for the implementer. 

From what I understand from the previous thread, it seems important to not interupt Step 3 "Process a render quantum." of the rendering loop, hence the microtask handling is moved to Step 4. 

So we would still have to spec when tasks are handled then.

On another point, I think we can also assume that the actual "audio backend" could be running on another thread, or even another process? I read something(sorry no link at hand) about Chromium moving Audio to a dedicated Mojo service? So that would mean the rendering thread would communicate the "render result" over ipc to the Audio service? 

In that case, could it perhaps be realistic to spec the rendering loop in the following way:

1. Let render result be false.
2. Process the control message queue.
3. Optionally, [run](https://html.spec.whatwg.org/multipage/#event-loop-processing-model) one, or several, steps of the event-loop of the AudioWorkletGlobalScope. (one step is running one task, and one microtask checkpoint).
4. Process a render quantum.
5. Communicate render result to the audio backend
6. Run a micro-task checkpoint for the event-loop of AudioWorkletGlobalScope.

The reasoning behind this list:

When you get to 5, you communicate with the audio backend. That means the backend will have to processs the result, and then communicate back via a control message to potentially ask for more data.

So while the backend is handling the result, it's a good time to do other work. 

The first thing you need to do is perform a microtask checkpoint, to resolve promises resulting from calls inside `process`. That's sort of necessary since otherwise by the next call to `process`, the processor will still not have seen the result from the promise, which might change it's state. 

Then, you go back to 2, and that's where you either have received a new control message, or not. If not, it's really a good time to run Step 3. Otherwise, if for example the backend sent a message asking for more data, the UA could skip Step 3. The important thing is that by putting the task handling at step 3, the UA already knows whether it has received a control message or not, and what the content of that message was. 
 
Also, the idea is really that you're not continuously running the event-loop of the worklet global scope, that is something that is explicitally "driven" by step 3 of the rendering loop, precluding any race-condition between an potential `onmessage` handler and the `process` call. 

Also, it's somewhat not spec compliant, since HTML says that an "An event loop must continually run through the following steps for as long as it exists". 

That's maybe a good point in the HTML spec to add, "unless it a [worklet-event-loop](https://html.spec.whatwg.org/multipage/webappapis.html#worklet-event-loop), in which case the UA can decide on per-case basis when to run steps of the event-loop(cc @annevk re https://github.com/whatwg/html/issues/4213)

**Comment by padenot** [comment]
In general, I agree with your two messages, this is something we're missing.

> From what I understand from the previous thread, it seems important to not interupt Step 3 "Process a render quantum." of the rendering loop, hence the microtask handling is moved to Step 4(versus interleaving them with calls to `process`).
>So we would still have to spec when tasks are handled then.

> On another point, I think we can also assume that the actual "audio backend" could be running on another thread, or even another process? I read something(sorry no link at hand) about Chromium moving Audio to a dedicated Mojo service? So that would mean the rendering thread would communicate the "render result" over ipc to the Audio service?

The audio system code is generally running on another process in modern browsers, yes (at least Firefox, for now just on Linux but this will change soon, and Chrome everywhere I believe, even before Mojo). It's never running on the same thread as something else, and it's often at higher priority.

> So one can wonder what use-case is most important for developpers. **Having fast and steady calls to `process`, with the ability to do non-blocking sends on the port back to the audio worklet node, or having steady messages incoming on the port**? I can imagine one could focus on maybe only using incoming messages for some initial setup, and then mostly using the port to send stuff back to the node on the control thread, while focusing on `process`, and not be sending messages back to the processor while audio is being processed(and off-course a call to `close` on the audio context would still be guaranteed to be handled as part of the control message queue).

Regarding the sentence I put in bold: it's preferable to do the former, otherwise it would break the fundamental rule of audio programming: there would be a risk of delaying the audio callback (see http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing for a more in depth explanation).

Things should be defined so it's clear, but a program that uses Promises or uses a lot of events as part of its audio processing is buggy. It's possible because they cannot be disabled. https://github.com/tc39/ecma262/issues/1120 has the background for this. The right thing to do is to use a `SharedArrayBuffer` for communication, and never ever create objects on this rendering thread.

The list of steps should be roughly:

1. get an event from the system that more audio should be played out
1. process the control messages
1. process the events, by getting all the events currently in the queue, and executing them as usual, with microtask checkpoints in between the events. The events that are posted during the rest of the steps are process next time this algorithm runs
1. do the audio processing, including calling `process` on the worklets
1. do a microtask checkpoint
1. return

We could even spec this as a normal event loop, if 2 and 4 are events, and 1 is an event that synchronously waits for a signal from the implementation that there is more audio to render.

**Comment by karlt** [comment]
There would be value in allowing the implementation to process control messages and events while waiting for a signal from the system that more audio should be processed.

**Comment by gterzian** [comment]
>  break the fundamental rule of audio programming: there would be a risk of delaying the audio callback (see http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing for a more in depth explanation).

Thanks, very interesting article and I think a good foundation on which a UA could make potential decisions with regards to prioritizing things.

> Things should be defined so it's clear, but a program that uses Promises or uses a lot of events as part of its audio processing is buggy. It's possible because they cannot be disabled.

In the context of the HTML [processing model](https://html.spec.whatwg.org/multipage/#event-loop-processing-model), a UA can actually throttle tasks of a given task-queue, or prioritize one task-queue over all others, as long as within a given task-queue, ordering is preserved. 

However, with microtasks it's a bit different, because if you run a task, you must run a microtask-checkpoint after it. 

So the question becomes: semantically, what is a call to `process` of a audio worklet processor? Is it a task or not? The fact that it runs in the context of a global scope would seem to imply it is.

And if it's a task, could you spec a specific "worklet event-loop" that would not perform microtask checkpoints after running each task? In theory it could be possible to queue all micro tasks on a dedicated task-queue, and run them as regular tasks, and run all tasks without a dedicated "microtask checkpoint".  That could then be a way to "ignore" micro-tasks, and fully prioritize a potential "rendering" task-queue(where a task would call `process`). 

At that point you could spec the "rendering loop" as a subtype of a "worklet loop", that doesn't do microtask checkpoints, which would give UA full flexibility to prioritize one task-queue over all others. 


>     1. get an event from the system that more audio should be played out
> 
>     2. process the control messages
> 
>     3. process the events, by getting all the events currently in the queue, and executing them as usual, with microtask checkpoints in between the events. The events that are posted during the rest of the steps are process next time this algorithm runs
> 
>     4. do the audio processing, including calling `process` on the worklets
> 
>     5. do a microtask checkpoint
> 
>     6. return

It could be easier to spec this as a formal event-loop, with the various steps being different task-queues that the UA could prioritize. 

Something like:

-  "control message" => "control-task-source"
-  "process the events" => all the the other task-sources, like the [port-message-queue](https://html.spec.whatwg.org/multipage/web-messaging.html#port-message-queue) of message ports. 
- "do the audio processing" => "render-a-quantum-task-source"? Enqueued itself by a task received from the  "control-task-source"?
- "do a microtask checkpoint" => Either do it consistently after running a task, or spec that worklet event-loops can treat that as a separate task-source and can skip the microtask checkpoint after running a task, in which case you have a "microtask-queue" that is just anoter task-queue that can potentially be throttled. 

> There would be value in allowing the implementation to process control messages and events while waiting for a signal from the system that more audio should be processed.

You could express that via giving full flexibility to UA in prioritizing task-sources, as opposed to via an imperative list of steps on the rendering loop. 

In practice, it could result in a sequence like:

1. You start a new iteration of the rendering loop.
2. The "render-a-quantum-task-source" is empty.
3. The "control-task-source" contains a task.
4. The "port-message-queue" contains a task too. 
5. For good measure, the "microtask-queue" contains several tasks. 
6. You select the task from the "control-task-source", it's a task from the backend asking for a render quantum. That queues a task on the "render-a-quantum-task-source".

7. You start another iteration of the loop. 
8. The "render-a-quantum-task-source" contains a task(and so does the "port-message-queue" and the "microtask-queue", in fact, those filled up even more). 
9. You select the task from the "render-a-quantum-task-source"  and run the "processing a render-quantum steps".

**Comment by gterzian** [comment]
By the way, forgive me for going on and on about this, I was just thinking about it and realized that worklet might be a first in terms of having to tightly integrate an HTML event-loop with it's surrounding environment, hence the spec is still lacking the right building blocks.

In practice when you run a window or worker loop, you handle plenty of "control messages" and don't run them as tasks, or interleave them with microtask checkpoints. That's because those messages aren't directly running JS. 

But with an audio worklet, you want the `process` method of the processor to be called as part of an "outer" system loop, yet those also run JS code. And because of the need of interop at the JS level, suddenly you need to spec that "outer" loop using HTML semantics. 

So you can either take a somewhat ad-hoc approach and mention a special "control-message-queue", and a "process a render quantum" step, which doesn't map to known concepts in HTML, or you can try to actually frame things in terms of tasks and task-queues, like I've tried above. 

The problem there is that running a "task", implies doing a "microtask checkpoint", which doesn't map to performance goals of calling `process` as part of a single "process a render quantum" operation that should be real-time. 

So I think we might really want to explore dropping the mandatory "microtask checkpoint" from worklet event-loops, so that we can still spec things using known HTML event-loop concepts, and still give implementations the ability to meet performance needs while staying spec compliant.

**Comment by gterzian** [comment]
> We could even spec this as a normal event loop, if 2 and 4 are events, and 1 is an event that synchronously waits for a signal from the implementation that there is more audio to render.

Yes I agree, and note again the problem that speccing something as a task, means that in theory it should be followed by a microtask checkpoint. Hence the need, I think, to introduce some flexibility for a dedicated "worklet event-loop" to not run microtask checkpoints after every task, instead treating it as just another task-queue. 

Re "1 is an event that synchronously waits for a signal from the implementation that there is more audio to render", you might want to merge that into 2, since something like a `close` control message coming from the audio context should probably be handled on par with the system asking for more data.

**Comment by gterzian** [comment]
By the way, sorry for writing looong posts on this, hoping it's still readable, and mainly trying to bounce off some ideas. 

On second thoughts, I came to realize my suggestion above would turn the "rendering loop" basically into the event-loop of the worklet global-scope, which might be taking things a bit too far in one direction. 

It might actually make sense to spec a "rendering loop" that isn't using HTML event-loop semantics, and rather reflects that realities of dealing with real-time audio(like is done now). 

The problem of the current spec is that it doesn't say when you should run the "tasks"(not the microtasks, which are mentioned) or the worklet global scope, for example for incoming messages on the message port of a processor. 

So a more "direct" solution to that problem, versus re-writing the entire loop, is what I initially proposed in my first post:  step 4 could process all task-queues of the event-loop of the AudioWorkletGlobalScope, with microtask checkpoints interleaved? (as opposed to only run microtasks, as is specced now). 

And then we'd have to accept that calling `process` isn't really a "task", altough the JS running it can enqueue (micro)tasks. 

It would probably be important for Step 4 to first do a microtask checkpoint, handling any microtasks enqueued as part of calls to `process`, and then handle tasks on the various queues of the worklet global-scope, interleaving each task with a micro-task checkpoint. 

That would be quite a big "step 4", and it would be the responsibility of the developer to ensure it doesn't fill up with too many tasks...

**Comment by padenot** [comment]
F2F summary:
- The audio rendering thread runs a normal event loop. Whenever a rendering quantum needs to happen, a task will is queued to the audio rendering thread, which will handle the rendering of the audio.

This means that `onmessage`s will be executed first, with microtasks checkpoint in between those. After this, the audio rendering task will be be executed, calling the `process` methods. After this, a microtask checkpoint is performed.

**Comment by gterzian** [comment]
> This means that onmessages will be executed first, with microtasks checkpoint in between those. After this, the audio rendering task will be be executed, calling the process methods. After this, a microtask checkpoint is performed.

Have you considered defining a new "render-an-audio-quantum" task-source? Then it could be said that the event-loop on the rendering thread has three task-sources(I think):

1. A [port-message-queue](https://html.spec.whatwg.org/multipage/web-messaging.html#port-message-queue)
2. A [control-message-queue](https://webaudio.github.io/web-audio-api/#control-message-queue)
3. A "render-an-audio-quantum", or similarly named, task-source. 

Then the UA could pick a runnable task from any task-queue at each iteration of the loop, and each task will be followed a microtask checkpoint, without having to define anything special in the audio spec.

**Comment by jackschaedler** [comment]
I don't think this reproduction scenario adds anything new to the conversation, but here it is for posterity. It's maybe helpful since it describes a concrete way that this issue affects developers out in the wild: https://jackschaedler.github.io/offline-audio-worklet-repro/
